### PR TITLE
When passing amounts to gateways, pass as a string with precision issues addressed

### PR DESCRIPTION
**Pull request by sc0ttkclark**
Fixes https://github.com/strangerstudios/paid-memberships-pro/issues/1922

### All Submissions:

* [x] Have you followed the [Contributing guideline](https://github.com/strangerstudios/paid-memberships-pro/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/strangerstudios/paid-memberships-pro/pulls) for the same update/change?

### Changes proposed in this Pull Request:

Resolves problems with float precision and prevent passing along faulty floats to the gateway APIs.

### How to test the changes in this Pull Request:

1. Test with a membership level priced at $19.90
2. Test with one-time
3. Test with recurring
4. Test with trial at $19.90
5. Test checkout for every gateway

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

> BUG FIX: Resolved problems with PHP float precision and prevent passing along faulty floats to the gateway APIs.

### CODE REVIEW

**Comment by sc0ttkclark** [comment]
I'm working on writing tests for the new functions introduced here, then I will set this PR as ready for review.

File: includes/functions.php [hunk 1]
```diff
@@ -3119,6 +3119,97 @@ function pmpro_round_price( $price, $currency = '' ) {
...
+
+	$amount = (float) $amount;
+
+	$currency_info = pmpro_get_currency( $currency );
+
+	$price_info = [
+		// The amount represented as a float.
+		'amount'        => $amount,
+		// The flat amount represent (example: 1.99 would be 199).
+		'amount_flat'   => 0,
```
Comment (dparker1005) [inline includes/functions.php:3167]:
```suggestion
		'amount_flat'   => 0,
		// The number of decimals for this currency.
		'decimals'        => (int) $currency_info['decimals'],
```

File: includes/functions.php [hunk 2]
```diff
@@ -3119,6 +3119,97 @@ function pmpro_round_price( $price, $currency = '' ) {
...
+			// The decimal part of the amount as a string (example: 1.99 would be 99, 1.00 would be 00).
+			'decimal_string' => '',
+		],
+		// The currency information.
+		'currency'      => $currency_info,
+	];
+
+	// Enforce integer.
+	$currency_info['decimals'] = (int) $currency_info['decimals'];
+
```
Comment (dparker1005) [inline includes/functions.php:3184]:
```suggestion
```

**Review by dparker1005 — APPROVED** [review]
Looks good, just a quick suggestion to immediately define the 'decimals' key in the $currency_info array

File: includes/functions.php (hunk 2 above)
Comment (sc0ttkclark) [reply includes/functions.php:3184]: Just checking here -- the decimals value is a string in many cases, enforcing as integer sets the expectation up for the below logic.

File: includes/functions.php (hunk 1 above)
Comment (sc0ttkclark) [reply includes/functions.php:3167]: What was the purpose of adding this here if we pass it in `$price_info['currency']` already?

**Comment by ideadude** [comment]
I took a look as well and tested some different checkout scenarios. All looks good.

RE the test code. Is it possible that the tests like this wouldn't work on systems that aren't having the same precision power. The test overrides the settings, but I thought this issue might also be related to PHP versions and other server settings.

`
/**
 * @covers ::pmpro_round_price()
 */
public function test_precision_issue_with_float() {
	$this->assertEquals( '19.899999999999999', json_encode( 19.9 ) );
	$this->assertEquals( '19.890000000000001', json_encode( 19.89 ) );
}
`

So json_encode(19.9) might not break on some systems?

Should we leave the tests out, or are there other ways to confirm that the new functions return the results we expect in certain situations?

**Comment by sc0ttkclark** [comment]
@ideadude The test itself is meant to set the settings to those that are the normal defaults for PHP regardless of configuration changes. That way, it can be sure it's testing and confirming based on the default PHP config that may be used across most sites. I haven't known many PHP configs to have those settings customized, however the note on PHP.net specifies why setting this manually here is most important:

From [PHP.net](https://www.php.net/manual/en/ini.core.php):

> Prior to PHP 7.1.0, the default value was 17.

So now [serialize_precision](https://www.php.net/manual/en/ini.core.php#ini.serialize-precision) is set to the value of `-1` by default in PHP 7.1+ so I'm just setting it to the most common denominator for testing purposes.

**Pull request merged**